pkg/rules/adapters/ingest.go

```diff
@@ -0,0 +1,97 @@
+package adapters
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/timescale/promscale/pkg/pgmodel/ingestor"
+	"github.com/timescale/promscale/pkg/pgmodel/metrics"
+	"github.com/timescale/promscale/pkg/prompb"
+)
+
+var samplesIngested = metrics.IngestorItems.With(map[string]string{"type": "metric", "kind": "sample", "subsystem": "ruling"})
+
+type ingestAdapter struct {
+	ingestor *ingestor.DBIngestor
+}
+
+func NewIngest(ingestor *ingestor.DBIngestor) *ingestAdapter {
+	return &ingestAdapter{ingestor}
+}
+
+type appenderAdapter struct {
+	// Map allows us to group similar samples into a single prompb.TimeSeries
+	// We can technically just maintain a list and have a single sample in each
+	// []prompb.TimeSeries like how Prometheus writes, but grouping samples gives
+	// us advantage in our ingestor.
+	data         map[uint64]*prompb.TimeSeries
+	ingestorCopy *ingestor.DBIngestor
+}
+
+func (a *ingestAdapter) Appender(_ context.Context) storage.Appender {
+	return &appenderAdapter{
+		data:         make(map[uint64]*prompb.TimeSeries),
+		ingestorCopy: a.ingestor,
+	}
+}
+
+func (app *appenderAdapter) Append(_ storage.SeriesRef, l labels.Labels, t int64, v float64) (storage.SeriesRef, error) {
+	hash := l.Hash()
+	if _, found := app.data[hash]; !found {
+		app.data[hash] = initTs(l)
+	}
+	app.data[hash].Samples = append(app.data[hash].Samples, prompb.Sample{Timestamp: t, Value: v})
+	return 0, nil // return 0 seriesRef since we do not do series level caching.
```

Comment:
No, because then we will need sync.Map. But that's not it, the rollback will be tough and we no longer know which samples belong to the current session, hence its much cleaner to keep a struct per session (appended). This is the same reason prometheus has written the code in that way.